### PR TITLE
[sourcekit] Add an optional virtual file system per request and/or open document

### PR DESCRIPTION
**Pull request by benlangmuir**
Enable sourcekitd requests to use a custom filesystem in place of (or on top of) the real filesystem. By default, an in-memory filesystem can be used with the `key.vfs.name: "in-memory-vfs"` option.  Vendors can also provide additional filesystems by implementing the `FileSystemProvider` interface and registering with `SwiftLangSupport::setFileSystemProvider`.  Requests can specify the desired filesystem using the `key.vfs.name` (string) and `key.vfs.options` (dictionary) keys.

The requests `codecomplete`, `codecomplete.open`, `codecomplete.update`, `editor.open` and `cursor_info` are supported, and it should be straightforward to add the remaining requests.  For one-off requests like `codecomplete`, the specified filesystem is used.  When using `editor.open`, the specified filesystem is associated with the given editor document. Future requests such as `editor.replacetext`, or `cursor_info` in the same document will continue to use that filesystem.  If `cursor_info` itself has a filesystem specified, it will be used instead of any associated with the editor document.  Currently there is no way to update the filesystem associated with an editor document other than re-opening the document.

---

This is an updated version of #24417 and much of the work is by @marcrasi 

Changes since #24417:
* Moved filesystem providers to live in the SwiftLangSupport layer, and instantiation into the requests
* Migrated configuration of the file system to use `OptionsDictionary`
* Renamed the builtin vfs from "testvfs" to "in-memory-vfs", since it's useful in its own right
* Added the ability to pass "in-memory-vfs" file contents as strings
* Moved the caching of the VFS from the ASTProducer onto the SwiftEditorDocument's semantic info. This provides better semantics for when to perform rebuilds
  * Open document now has a single entry in the ASTCache that gets rebuilt instead of one per VFS
  * VFS identity is not considered in `shouldRebuild`, only the contents and timestamps
* Added additional tests for these state and caching changes

### CODE REVIEW

**Comment by benlangmuir** [comment]
@swift-ci please smoke test

**Comment by benlangmuir** [comment]
@swift-ci please smoke test macOS

**Comment by benlangmuir** [comment]
@swift-ci please smoke test

**Comment by benlangmuir** [comment]
@swift-ci please clean smoke test macOS

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test

**Review by marcrasi — COMMENTED** [review]
Thanks for helping with this!

This approach looks good to me, based on reading it. I'll try getting this patch to work with my client today, to see if there are any unexpected difficulties.

File: tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp
Comment (benlangmuir) [reply tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp]: I guess so, yeah.

File: tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp
Comment (benlangmuir) [reply tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp]: You have the right idea for how it works, but if you want to test it with cursor_info you need to be aware that cursor_info can also use the `canUseASTWithSnapshots` API, so even if the files have changed we might not rebuild the AST (which is no different than the current behaviour of cursor_info if the files on disk changed for example).

File: tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp
Comment (benlangmuir) [reply tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp]: I am so sick of our inconsistent style for names 😭  I'll... think about how to handle this, but I'd rather user lowerCamelCase for new code.

**Comment by benlangmuir** [comment]
> I'll try getting this patch to work with my client today, to see if there are any unexpected difficulties.

Let me know how it does!

File: tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp
Comment (marcrasi) [reply tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp]: ok!

**Review by marcrasi — APPROVED** [review]
I hooked my client up to this and it passes all the tests I have for my client and also the client seems to work when I play with it manually.

**Comment by marcrasi** [comment]
Are you waiting for anything else before you merge this in? Any way I can help?

**Comment by benlangmuir** [comment]
> Are you waiting for anything else before you merge this in? Any way I can help?

Nope, I just got busy with something else before I could apply your suggested changes.  I was hoping to get back to it this week, but if you want to take it back over, feel free :-)

**Comment by marcrasi** [comment]
I'll have some time to rebase this and update it today. So I'll do that today, unless you've already started working on it by now.

**Comment by benlangmuir** [comment]
I wouldn't get to it today, so go for it!

**Comment by marcrasi** [comment]
Rebased and updated, in https://github.com/apple/swift/pull/26168.

**Pull request closed without merging**
